### PR TITLE
Plushie relationships!

### PR DESCRIPTION
**Pull request by ghost**
:cl: CosmicScientist
add: You can make plushies kiss one another!
/:cl:

Missed opportunity/inspiration:
#32615
#32918
@MoreRobustThanYou   ❤️ 
@optimumtact I did what you said; sought approval from myself and here I lay.

We require more plushies and this is how barring some tribble trouble!

Also corrected one smidge of grammar that triggered me so terribly.

plushie relationship
Who the babby inherits type from should be a 50/50.

I would like to thank MSO and another coderbus denizen, whose name unfortunately escapes me, for making this possible.

### CODE REVIEW

**Comment by Kawaii-Big-Boss** [comment]
cease

**Comment by TDcoolguy300** [comment]
:thinking:

**Comment by ninjanomnom** [comment]
I regret making the plush type path

**Comment by ZeWaka** [comment]
I'm really excited for this PR to be merged! I think it adds a robust and dynamic relationship system that could definitely be expanded upon in the future. However, I'm concerned that there is no way to up my plushie's stats. I feel like enlisting my plushie in military service could up its' sense of duty, or even perhaps it's loyalty value. However, I really do enjoy this feature and I expect it to be merged soon.

**Comment by FantasticFwoosh** [comment]
Player references, cringe interactions, shipping. Ew.

**Comment by Hatterhat** [comment]
itt: cosmic goes fucking insane

holy shit zewaka this isnt goon what are you doing here

**Comment by CitrusGender** [comment]
cosmic I love you bud but you gotta think of the children

**Review by optimumtact — DISMISSED** [review]

File: code/game/objects/items/plushes.dm [hunk 1]
```diff
@@ -9,11 +9,52 @@
...
+	var/list/vowbroken_message
+	var/list/child_message
+	var/normal_desc
+	//--end of love :'(--
 
 /obj/item/toy/plush/Initialize()
 	. = ..()
 	AddComponent(/datum/component/squeak, squeak_override)
 
+/obj/item/toy/plush/New()
```
Comment (optimumtact) [inline code/game/objects/items/plushes.dm]: put this in intialise, not new

File: code/game/objects/items/plushes.dm [hunk 2]
```diff
@@ -9,11 +9,52 @@
...
+	normal_desc = desc
+
+	love_message += ("\n\The [src] is so happy, \he could rip a seam!")
+	heartbroken_message += ("\n\The [src] looks so sad.")
+	vowbroken_message += ("\n\The [src] lost \his ring...")
+	child_message += ("\n\The [src] can't remember what sleep is.")
+
+	return ..()
+
 /obj/item/toy/plush/Destroy()
```
Comment (optimumtact) [inline code/game/objects/items/plushes.dm]: null all the vars you added at this stage, and if they have a lover, null it's lover var pointing back to this object best if you have an unset_lover proc, so you can just lover.unset_lover(src)

Comment (optimumtact) [reply code/game/objects/items/plushes.dm]: also use the QDEL_LIST operator to clear the scorned list

File: code/game/objects/items/plushes.dm [hunk 3]
```diff
@@ -66,8 +107,143 @@
...
+/obj/item/toy/plush/proc/love(obj/item/toy/plush/Kisser, mob/living/user)	//~<3
+	var/chance = 100	//to steal a kiss, surely there's a 100% chance no-one would reject a plush such as I?
+	var/concern = 20	//perhaps something might cloud true love with doubt
+	var/loyalty = 30	//why should another get between us?
+	var/duty = 50		//conquering another's is what I live for
+
+	//we are not catholic
+	if (young == TRUE || Kisser.young == TRUE)
+		user.show_message("<span class='notice'>[src] plays tag with [Kisser].</span>", 1,
+			"<span class='notice'>They're happy.</span>", 0)
```
Comment (RemieRichards) [inline code/game/objects/items/plushes.dm]: *single tear*

File: code/game/objects/items/plushes.dm [hunk 4]
```diff
@@ -66,8 +107,143 @@
...
+
+	//then comes marriage
+	else if(Kisser.lover == src && Kisser.partner != src)	//need to be lovers (assumes loving is a two way street) but not married (also assumes similar)
+		user.visible_message("<span class='notice'>[user] pronounces \the [Kisser] and \the [src] married! D'aw.</span>",
+									"<span class='notice'>You pronounce \the [Kisser] and \the [src] married!</span>")
+		new_partner(Kisser)
+		Kisser.new_partner(src)
+
+	//then comes a baby in a baby's carriage, or an adoption in an adoption's orphanage
+	else if(Kisser.partner == src && !src.plush_child)	//the one advancing does not take ownership of the child and we have a one child policy in the toyshop
```
Comment (optimumtact) [inline code/game/objects/items/plushes.dm]: you shouldn't need src for src.plush_child

File: code/game/objects/items/plushes.dm [hunk 5]
```diff
@@ -9,11 +9,52 @@
 	var/list/squeak_override //Weighted list; If you want your plush to have different squeak sounds use this
 	var/stuffed = TRUE //If the plushie has stuffing in it
 	var/obj/item/grenade/grenade //You can remove the stuffing from a plushie and add a grenade to it for *nefarious uses*
+	//--love ~<3--
+	var/male_chance = 1
```
Comment (optimumtact) [inline code/game/objects/items/plushes.dm]: this should be a single var, gender, where 1 is male, 0 is female and 2 is randomly choose on INITIZIALIZE the gender.

use defines PUSH_MALE, PLUSH_FEMALE, PLUSH_RANDOM

Comment (Grazyn) [reply code/game/objects/items/plushes.dm]: >binary genders

smh

**Comment by Alek2ander** [comment]
The madman literally added ERP
:+1:

**Comment by nicbn** [comment]
I'm really scared

**Comment by Qbopper** [comment]
i can't stop laughing at the list of people who -1'd the OP

**Comment by pantherpusher** [comment]
How long until action fiogures can do this too?

**Comment by WJohn** [comment]
absolute madman

File: code/game/objects/items/plushes.dm [hunk 6]
```diff
@@ -66,8 +107,143 @@
...
+		user.show_message("<span class='notice'>[src] rejects the advances of [Kisser]!</span>", 1,
+			"<span class='notice'>That didn't feel like it worked.</span>", 0)
+	else if(src in Kisser.scorned)
+		user.show_message("<span class='notice'>[Kisser] realises who [src] is and turns away.</span>", 1,
+			"<span class='notice'>That didn't feel like it worked.</span>", 0)
+
+	//first comes love
+	else if(Kisser.lover != src && Kisser.partner != src)	//cannot be lovers or married
+		if(Kisser.lover)	//if the initiator has a lover
+			Kisser.lover.heartbreak(Kisser)	//the old lover can get over the kiss-and-run whilst the kisser has some fun
```
Comment (JJRcop) [inline code/game/objects/items/plushes.dm]: Shouldn't this only happen if the kiss succeeds?

File: code/game/objects/items/plushes.dm [hunk 7]
```diff
@@ -66,8 +107,143 @@
...
+			chance -= concern	//one heart already broken, what does another mean?
+		if(lover)	//if the recipient has a lover
+			chance -= loyalty	//mustn't... but those lips
+		if(partner)	//if the recipient has a partner
+			chance -= duty	//do we mate for life?
+
+		if(prob(chance))	//did we bag a date?
+			user.visible_message("<span class='notice'>[user] makes \the [Kisser] kiss \the [src]!</span>",
```
Comment (JJRcop) [inline code/game/objects/items/plushes.dm]: `\the`se should be removed.

File: code/game/objects/items/plushes.dm [hunk 8]
```diff
@@ -66,8 +107,143 @@
...
+				lover.heartbreak(src)
+			new_lover(Kisser)
+			Kisser.new_lover(src)
+		else
+			user.show_message("<span class='notice'>[src] rejects the advances of [Kisser], maybe next time?</span>", 1,
+								"<span class='notice'>That didn't feel like it worked, this time.</span>", 0)
+
+	//then comes marriage
+	else if(Kisser.lover == src && Kisser.partner != src)	//need to be lovers (assumes loving is a two way street) but not married (also assumes similar)
+		user.visible_message("<span class='notice'>[user] pronounces \the [Kisser] and \the [src] married! D'aw.</span>",
```
Comment (JJRcop) [inline code/game/objects/items/plushes.dm]: All of `\the`m.

File: code/game/objects/items/plushes.dm (hunk 7 above)
Comment (JJRcop) [inline code/game/objects/items/plushes.dm]: `\the`se should be removed, all of `\the`m. `\the`y're unnecessary.
It's ra`\the`r strange that some of `\the`se don't have 'em, but most o`\the`rs do.

**Comment by imsxz** [comment]
Your coders were so preoccupied with whether or not they could, they didn’t stop to think if they should.

**Comment by jughu** [comment]
Time to add a oranges plushy

**Comment by Xhuis** [comment]
I have no idea why this exists, but as long as Plushvar and Nar-Plush will always hate each other, I'm good.

**Comment by davethwave** [comment]
Make this a valentines event perhaps? Just my opinion but I don't think this would fit into the game very well by itself despite the work put into it.

**Comment by Hatterhat** [comment]
what if plushvar and narplush could hatefuck

File: code/game/objects/items/plushes.dm (hunk 3 above)
Comment (nicbn) [reply code/game/objects/items/plushes.dm]: Really, please remove this. This will attract the worst possible kind of players to our servers.

**Comment by Spaceinaba** [comment]
I was extremely against this PR, but OP appears to be a genius. That earns a solid 👍  from me

**Comment by nicbn** [comment]
Man the kissing is acceptable but everything else than it is pushing the line. ERP is not allowed currently.

**Comment by QualityVan** [comment]
There was a point where this shit had to stop and we've clearly passed it.

<s>Let's keep going and see what happens.

**Comment by qbmax32** [comment]
We all have fetishes
Cosmic is just very outspoken about his

File: code/game/objects/items/plushes.dm (hunk 1 above)
Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: Okay will do. Can I ask why so I will hopefully remember in future?

File: code/game/objects/items/plushes.dm (hunk 2 above)
Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: Good point. Will do!

File: code/game/objects/items/plushes.dm (hunk 4 above)
Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: Shhh I totally left that in to make sure maintainers are doing their job and totally didn't get told this by MSO in coderbus before I typed it in.

Here's your gold star 🌟

File: code/game/objects/items/plushes.dm (hunk 5 above)
Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: I will do that. I'm not sure my PR's scope requires specific gender weighting until there is a catgrille plush to simulate bagil population.

**Comment by juanototo2** [comment]
The emojicracy demands this be merged


## **MERGE IT**

File: code/game/objects/items/plushes.dm (hunk 7 above)
Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: _**Griff Murderbone** makes ratvar plushie kiss narsie plushie_
_**Griff Murderbone** makes the ratvar plushie kiss the narsie plushie_

I'll look at which ones should have `\the` parts removed or added. Would it be better to rename the toys? Having them all be `[something] plushie` takes up most word space.

**Comment by CosmicScientist** [comment]
@ZeWaka if intent based plushie interactions isn't prohibited and I can manage to code it, we can give them stats.

@CitrusGender I did! Did you not read the code comments? They just play tag.

@nicbn it's going to be okay.

@pantherpusher that was my original intention before the code told me plushies were not strictly the same, so @ninjanomnom shouldn't be too upset. I can, good boy points permitting, create a PR after this which expands it to more toys!

@imsxz for some reason I'm thinking of reactions to goof. Does this mean I'm pro-goofcode now?

@Xhuis I'll make sure to test it properly so they don't both try to execute.

@Hatterhat hmmm...

**Comment by Buggy123** [comment]
wat

**Comment by pantherpusher** [comment]
"You make Plushvar and Narplush kiss. They don't seem too thrilled about it."

**Comment by mattroks101** [comment]
This would make me come back and play TG again.

File: code/game/objects/items/plushes.dm (hunk 3 above)
Comment (optimumtact) [reply code/game/objects/items/plushes.dm]: I can't tell if this is sarcasm or not, help me out

**Comment by optimumtact** [comment]
Certainly a lot of people worried about if a silly interaction on a plushie that might show up three times in a round might 'fit' in our game that has actual catpeople in it.

:thinking: :thought_balloon:

File: code/game/objects/items/plushes.dm (hunk 5 above)
Comment (Cyberboss) [reply code/game/objects/items/plushes.dm]: Just use the built in BYOND defines for gender and have NEUTRAL be random

File: code/game/objects/items/plushes.dm (hunk 1 above)
Comment (optimumtact) [reply code/game/objects/items/plushes.dm]: We have better control over when the Initialize proc runs, so we can better smooth out the creation of objects and manage it with our ticker subsystem, as opposed to new, which just runs straight away.

File: code/game/objects/items/plushes.dm (hunk 5 above)
Comment (optimumtact) [reply code/game/objects/items/plushes.dm]: yeah okay that will work fine as well

**Comment by Okand37** [comment]
Victory costs. Every time, you pay a little more.

**Comment by Bawhoppen** [comment]
Freedom isn't free

**Comment by PKPenguin321** [comment]
allows you to build a plushie baby engine by way of dumping plush babies onto a conveyor belt that launches into a supermatter
👍

**Comment by CosmicScientist** [comment]
Changes as per request:
* moved New() code to Init()
* performed cleanup in Destroy()
* removed src.
* simplified gender

**Comment by Xhuis** [comment]
This is rapidly becoming the best PR of 2017.

**Comment by CosmicScientist** [comment]
to do (PR should otherwise be merge ready):
- [ ] remove `\the` and adjust plushie naming convention to minimise repetition
- [ ] make use of the lists for moods so a slime isn't wearing a ring on its non-existent finger
- [ ] test Plushvar and Narplush
- [ ] consider if the code or the toy owner should decide if Plushvar and Narplush can overcome society's rules and find out they loved one another all along
- [ ] finish off with a note on intended future features for another PR (e.g. extend to all toys, intent based interactions, ~~pokemon~~ plush RPG stats, scare nicbn again)

File: code/game/objects/items/plushes.dm [hunk 9]
```diff
@@ -9,13 +9,88 @@
...
+	normal_desc = desc
+
 /obj/item/toy/plush/Destroy()
 	QDEL_NULL(grenade)
+
+	//inform next of kin and... acquaintances
+	if(partner)
+		partner.bad_news(src)
+		partner = null
+		lover = null
```
Comment (optimumtact) [inline code/game/objects/items/plushes.dm]: you're nulling lover before trying to reference it below

Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: The else if?

Comment (optimumtact) [reply code/game/objects/items/plushes.dm]: oh ignore me, didn't see the elseif

File: code/game/objects/items/plushes.dm [hunk 10]
```diff
@@ -66,8 +141,196 @@
...
+/obj/item/toy/plush/proc/plop(obj/item/toy/plush/Daddy)
+	if(partner != Daddy)
+		return	//we do not have bastards in our toyshop
+
+	if(prob(50))	//it has my eyes
+		plush_child = new type(get_turf(loc))
+	else	//it has your eyes
+		plush_child = new Daddy.type(get_turf(loc))
+
+	plush_child.make_young(src, Daddy)
```
Comment (optimumtact) [inline code/game/objects/items/plushes.dm]: set the mood message of the parents in the make_young proc rather than outside it

Comment (optimumtact) [reply code/game/objects/items/plushes.dm]: i.e move the below code into the make young proc for each parent

**Comment by CosmicScientist** [comment]
Additional personal to do:
- [ ] ❤Nuclear family❤

File: code/game/objects/items/plushes.dm (hunk 10 above)
Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: Good catch.

**Comment by Grazyn** [comment]
Is consaguinity accounted for, to prevent sibling from marrying?

**Comment by DoctoreMad** [comment]
>33 thumbs up

Why

**Comment by imsxz** [comment]
@grazyn hey man if two plushies love each other very much we shouldn't try and stop them

**Comment by DoctoreMad** [comment]
@Imsxz 
59f3d200f1462e32cc3cc31a41fadbf469702eb12405abbf94b280a2d2157a20

**Comment by nicbn** [comment]
I just want this to end

**Comment by nicbn** [comment]
What have tg coders became

**Comment by Xhuis** [comment]
The server isn't going to fall apart if cutesy plushie relationships get added, nicbn. You are acting like a child.

**Comment by nicbn** [comment]
ERP stills ERP even if it's not between players. What are we saying to our players by admitting this kind of feature? That we don't have a solid rule enforcement?

**Comment by Xhuis** [comment]
That having cutesy plushie relationships is okay? What the hell kind of drugs are you *on?*

**Comment by nicbn** [comment]
It's not cute it's gross
argh

File: code/game/objects/items/plushes.dm (hunk 7 above)
Comment (JJRcop) [reply code/game/objects/items/plushes.dm]: `\the [Kisser]` doesn't do anything different than `[Kisser]`. If Kisser has a proper name, then `\the` doesn't do anything and is omitted. If kisser has an improper name, `\the` is included. this is what you expect, but what you don't know is that just `[Kisser]` does the exact same thing even without `\the`, so we don't use `\the` anymore.

Comment (JJRcop) [reply code/game/objects/items/plushes.dm]: `\the [Kisser]` doesn't do anything different than `[Kisser]`. If Kisser has a proper name, then `\the` doesn't do anything and is omitted. If kisser has an improper name, `\the` is included. This is what you expect, however, `[Kisser]` on its own does the same thing even without `\the`, so we don't use `\the` anymore.

**Comment by JJRcop** [comment]
I commented on how `\the` is not necessary [here](https://github.com/tgstation/tgstation/pull/33184#discussion_r154411458).

File: code/game/objects/items/plushes.dm (hunk 7 above)
Comment (CosmicScientist) [reply code/game/objects/items/plushes.dm]: Yeah I found out that after testing. Sorry for not realising this was filtered for in the first place.

The `\the`s will be gone in the next commit, I'm just flailing at a feature breaking bug right now.

**Comment by CosmicScientist** [comment]
@Grazyn blood relations cannot be together as children are young and not allowed an adult relationship. The only way to permit keeping it in the family is var editing young away and I can't imagine any proper admin would perform such an act to a poor, defenceless plush.

**Comment by CosmicScientist** [comment]
@nicbn let your hair down, stop getting your knickers in a twist and have some fun. Or don't use the feature.

**Comment by CosmicScientist** [comment]
@DoctoreMad maybe admins should have a turn into salt pillar button?

**Comment by Bawhoppen** [comment]
How are people even comparing this to ERP

**Comment by nicbn** [comment]
Anything more sensualized than kisses is ERP and blahblahblah. Having the kissing and mariage is ok, but not the... reproduction thing.

**Comment by Bawhoppen** [comment]
But what erotic stuff is actually happening?

**Comment by Pascal125** [comment]
When mama and daddy plushie love eachother very much. Poof. A baby is appear.

Apparently.

**Comment by CosmicScientist** [comment]
- all instances of `\the` removed
- parental mood shift moved to make_young()
- small changes here and there including changing the child to inherit the name of the owning parent + Jr since the generic name would make it confusing (as there can only be a second generation in this PR's scope, no more future proofing is required), especially if the parent's mood message is changed, example below:
https://i.imgur.com/mqjY5EM.png

**Comment by Ausops** [comment]
this is wrong on so many levels

advanced plushie/toy functionality means many things but it shouldn't have meant reproduction

**Comment by Xhuis** [comment]
Okay, Cosmic, *first* of all, how dare you.

**Comment by ChangelingRain** [comment]
I'm finding this more amusing than I should. It lines up so well, really.

**Comment by Xhuis** [comment]
I can't believe you're trying to perpetuate this insanity.

**Comment by DoctoreMad** [comment]
@optimumtact A lotta loyalty for a meme PR

**Comment by praisenarsie** [comment]
its over

**Comment by davethwave** [comment]
The next few days will be interesting...

**Comment by CosmicScientist** [comment]
Not quite over, as I need to finish this off with future intended features and modifications that were outside the scope of getting it working:
- [ ] refactor this file into multiple as I feel 507 lines is overwhelming, they could be split into plushies, clash of the plushies, plushie love and assorted procs with the base plush, or as coderbus suggests, one char per file
- [ ] kustom mood text for each plush
- [ ] lists of kustom names for the plushies to draw from on creation (so it's not just slime plushie and slime plushie Jr)
- [ ] ❤Nuclear family❤
- [ ] intent based interactions
- [ ] plushie stats to hone, such as huggability, throwforce and how adorbs it is

**Comment by CosmicScientist** [comment]
I would like to thank oranges, MSO and coderbus at large for the help they gave me because I sucked so badly I didn't know to initialise a list before populating it. Or that there was a runtime log.

So, thank you, you wonderful people.

**Comment by calzilla1** [comment]
Cosmic, are you.a furry?

**Comment by nicbn** [comment]
Some claim we're reaching byond limits
And that's what we are wasting our precious code limits on

**Comment by adamkad1** [comment]
You should make special combos for some plushies like ratvar and narsie

**Comment by Rowelll** [comment]
1 year ago

**Comment by ike709** [comment]
@adamkad1 Love how you commented on a PR that's over one year old to request a feature that has already existed longer this PR.

**Comment by adamkad1** [comment]
oh god i once again forgot to look at the date, curse me

**Pull request merged**